### ISSUE TITLE
Write tests to the application

### ISSUE THREAD

**Issue by kerollaine** [sigalei/sigalei-stack#26]
There's a lot of information about how important it is to write tests to an application. I suggest starting this with unit tests.
Unit tests can ensure that our application keeps working right after changes in the code base, as well it's a way to provide documentation for what the code is actually meant to do, what is great when new people join us, for example.

Tests can also improve productivity because automation allow run tests easily, in seconds otherwise spend the time testing the application manually. The side effects of changes can be more easily identified. Moreover, tests is a huge step to improve the CI/CD pipeline.

This article [here](https://martinfowler.com/articles/practical-test-pyramid.html) explain with details more about the tests concepts, type of software testing, etc. I think it will helps understand the reasons why we should write tests.

**Comment by caioflores** [comment]
We totally agree with you! It's a great suggestion! We already started to write some unit tests in our spiders in python, but in our application we just explored some libraries and it's not working yet.

Do you have experience with unit tests? Did you feel this benefits in any project that you worked?

**Comment by kerollaine** [comment]
I think it to test spiders can be more harder than usual. You use Scrapy framework? Spiders Contracts can be useful.

I already worked with unit tests in Python projects and currently I use RSpec or Minitest with Ruby to do this. I believe the whole team benefits from this every day. Having tests is especially important to see how a new change can affect the project, as well test deploys and avoiding problems and bugs in production environment. In some cases other test types are used as integration tests per example.
When I join a new team, tests make possible understand the source code, the business and patterns of the project much more quickly, even when I still very little knowledge about the code. 

Sometimes is very nice develop a new feature using TDD, because this has power to change the final code a lot - to the best. I liked to use TDD when I pair programming because I notice a better understanding of both developers about the problem to be solved. 

In a nutshell, I can see many gains for us having more tests allied to a clean code on our project. At first, it may seem so much work but will worth it in the near future and anyone who maintains the code will do this much more easily and more safely.

**Comment by tretinha** [comment]
Yup. We use the Scrapy framework, in our case contracts weren't very useful because of the high change rate of the pages, which eventually led to our spiders failing a few tests out of nowhere. Currently, we are using an in-house solution for testing spiders. If you want to know more of how we are doing this, feel free to chat with me in the slack channels (opensanca, react-brasil, etc.)

**Comment by kerollaine** [comment]
Interesting @tretinha, I want to know more about it.. I will looking for you in OpenSanca.

**Comment by kerollaine** [comment]
I'm working on a initial test using `easygraphql-tester`
